morango/syncsession.py

```diff
@@ -21,21 +33,157 @@ class Connection(object):
+                "remote_instance": session_resp.data.get("instance") or "{}",
+            }
+            self.sync_session = SyncSession.objects.create(**data)
+
+            return SyncClient(self)
+
+    def get_remote_certificates(self, primary_partition):
+        remote_certs = []
+        # request certs for this primary partition, where the server also has a private key for
+        remote_certs_resp = self._request(api_urls.CERTIFICATE, params={'primary_partition': primary_partition})
+
+        # inflate remote certs into a list of unsaved models
+        for cert in json.loads(remote_certs_resp.data):
+            remote_certs.append(Certificate.deserialize(cert["serialized"], cert["signature"]))
+        return remote_certs
+
+    def certificate_signing_request(self, parent_cert, scope_definition_id, scope_params, username=None, password=None):
+        csr_key = Key()
+        # build up data for csr
+        data = {
+            "parent": parent_cert.id,
+            "profile": self.profile,
+            "scope_definition": scope_definition_id,
+            "scope_version": parent_cert.scope_version,
+            "scope_params": scope_params,
+            "public_key": csr_key.get_public_key_string()
+        }
+        csr_resp = self._request(api_urls.CERTIFICATE, method="POST", data=data, auth=(username, password))
+        csr_data = json.loads(csr_resp.data)
+
+        # verify cert returned from server, and proceed to save into our records
+        csr_cert = Certificate.deserialize(csr_data["serialized"], csr_data["signature"])
+        csr_cert.private_key = csr_key
+        csr_cert.check_certificate()
+        csr_cert.save()
+        return csr_cert
 
-    def request_certificate(self, unsaved_cert, parent_cert, username=None, password=None):
         pass
 
-    def _push_chunk(self, data):
-        raise NotImplementedError
+    def _get_certificate_chain(self, server_cert):
+        # get ancestors certificate chain for this server cert
+        cert_chain_resp = self._request(api_urls.CERTIFICATE, params={'ancestors_of': server_cert.id})
+
+        # upon receiving cert chain from server, we attempt to save the chain into our records
+        Certificate.save_certificate_chain(cert_chain_resp.data, expected_last_id=server_cert.id)
+
+    def _create_transfer_session(self, push, filter_prefixes):
+        # build data for creating transfer session on server side
+        data = {
+            'id': uuid.uuid4().hex,
+            'filter': filter_prefixes,
+            'push': push,
+            'sync_session_id': self.sync_session.id,
+            'local_fsic': DatabaseMaxCounter.calculate_filter_max_counters(filter_prefixes)
+        }
+        transfer_resp = self._request(api_urls.TRANSFERSESSION, method="POST", data=data)
+
+        # if transfersession successfully created, create on our own side
+        if transfer_resp.status_code == status.HTTP_201_CREATED:
```

Comment:
Probably want to raise something if it fails, also.

Reply:
And I think we decided it would be useful to save the transfersession locally before doing the post, since for a PULL it might take quite a while to respond (as it queues things up on the server side), and it's nice to have the transfersession object existing locally to show status while that happens.